File: machine/cpp/machine/devices/audio/command_executor.cpp
#include "machine/devices/audio/command_executor.h"

#include "spec/bmsx/io.h"
#include "machine/common/numeric.h"
#include "machine/devices/audio/active_slots.h"
#include "machine/devices/audio/command_fifo.h"
#include "machine/devices/audio/output.h"
#include "machine/devices/audio/sample_memory.h"
#include "machine/devices/audio/selected_slot_latch.h"
#include "machine/devices/audio/service_clock.h"
#include "machine/devices/audio/source.h"
#include "machine/devices/audio/slot_bank.h"
#include "machine/devices/device_status.h"
#include "machine/memory/memory.h"
#include "machine/scheduler/device.h"

namespace bmsx {

ApuCommandExecutor::ApuCommandExecutor(Memory& memory,
	ApuOutputMixer& audioOutput,
	DeviceScheduler& scheduler,
	ApuCommandFifo& commandFifo,
	ApuSampleMemory& sampleMemory,
	ApuActiveSlots& activeSlots,
	ApuSlotBank& slots,
	ApuSelectedSlotLatch& selectedSlotLatch,
	DeviceStatusLatch& fault,
	ApuServiceClock& serviceClock,
	const ApuParameterRegisterWords& commandRegisterWords)
	: m_memory(memory)
	, m_audioOutput(audioOutput)
	, m_scheduler(scheduler)
	, m_commandFifo(commandFifo)
	, m_sampleMemory(sampleMemory)
	, m_activeSlots(activeSlots)
	, m_slots(slots)
	, m_selectedSlotLatch(selectedSlotLatch)
	, m_fault(fault)
	, m_serviceClock(serviceClock)
	, m_commandRegisterWords(commandRegisterWords) {}

void ApuCommandExecutor::drainCommandFifo() {
	while (!m_commandFifo.empty()) {
		const u32 command = m_commandFifo.popInto(m_commandDispatchRegisterWords);
		executeCommand(command, m_commandDispatchRegisterWords);
	}
}

void ApuCommandExecutor::restoreOutputVoice(const ApuOutputVoiceState& state) {
	const ApuAudioSlot slot = state.slot;
	m_slots.loadRegisterWords(slot, m_slotRegisterDispatchWords);
	loadApuAudioSource(m_source, m_slotRegisterDispatchWords);
	bindSource(m_source, state.sourceCartridgeSlot);
	m_audioOutput.restoreVoice(
		slot,
		m_source,
		m_sourceView,
		m_slotRegisterDispatchWords,
		state
	);
}

u32 ApuCommandExecutor::selectedSlotRegisterReadThunk(void* context, u32 addr, MappedBusSignals) {
	auto& executor = *static_cast<ApuCommandExecutor*>(context);
	const i64 nowCycles = executor.m_scheduler.currentNowCycles();
	executor.m_serviceClock.synchronize(nowCycles);
	const u32 slot = executor.m_commandRegisterWords[APU_PARAMETER_SLOT_INDEX] & APU_SLOT_INDEX_MASK;
	const u32 parameterIndex = (addr - IO_APU_SELECTED_SLOT_REG0) / IO_WORD_SIZE;
	return executor.m_slots.registerWord(slot, parameterIndex);
}

void ApuCommandExecutor::selectedSlotRegisterWriteThunk(void* context, u32 addr, u32 value, MappedBusSignals) {
	auto& executor = *static_cast<ApuCommandExecutor*>(context);
	const i64 nowCycles = executor.m_scheduler.currentNowCycles();
	executor.m_serviceClock.synchronize(nowCycles);
	const u32 slot = executor.m_commandRegisterWords[APU_PARAMETER_SLOT_INDEX] & APU_SLOT_INDEX_MASK;
	executor.writeSlotRegisterWord(slot, (addr - IO_APU_SELECTED_SLOT_REG0) / IO_WORD_SIZE, value);
	executor.m_serviceClock.scheduleNext(nowCycles);
}

void ApuCommandExecutor::executeCommand(u32 command, const ApuParameterRegisterWords& registerWords) {
	switch (command) {
		case APU_CMD_PLAY:
			play(registerWords);
			return;
		case APU_CMD_STOP_SLOT:
			stopSlot(registerWords);
			return;
		case APU_CMD_SET_SLOT_GAIN:
			setSlotGain(registerWords);
			return;
		default:
			m_fault.raise(APU_FAULT_BAD_CMD, command);
			return;
	}
}

void ApuCommandExecutor::play(const ApuParameterRegisterWords& registerWords) {
	loadApuAudioSource(m_source, registerWords);
	const ApuAudioSlot slot = registerWords[APU_PARAMETER_SLOT_INDEX] & APU_SLOT_INDEX_MASK;
	startPlay(m_source, slot, registerWords);
}

void ApuCommandExecutor::startPlay(const ApuAudioSource& source, ApuAudioSlot slot, const ApuParameterRegisterWords& registerWords) {
	if (!bindSource(source, m_memory.cartridgeController().selectedSlot())) {
		m_fault.raise(APU_FAULT_SOURCE_RANGE, source.sourceAddr);
		return;
	}
	m_activeSlots.setActive(slot, registerWords);
	m_audioOutput.playVoice(slot, source, m_sourceView, registerWords);
}

void ApuCommandExecutor::stopSlot(const ApuParameterRegisterWords& registerWords) {
	const ApuAudioSlot slot = registerWords[APU_PARAMETER_SLOT_INDEX] & APU_SLOT_INDEX_MASK;
	const u32 fadeSamples = registerWords[APU_PARAMETER_FADE_SAMPLES_INDEX];
	if ((m_slots.activeMask() & (1u << slot)) == 0u) {
		m_audioOutput.stopSlot(slot);
		m_activeSlots.stop(slot);
		return;
	}
	if (fadeSamples > 0u) {
		m_activeSlots.setPhase(slot, APU_SLOT_PHASE_FADING);
		m_audioOutput.stopSlot(slot, fadeSamples);
		return;
	}
	m_audioOutput.stopSlot(slot);
	m_activeSlots.stop(slot);
}

void ApuCommandExecutor::setSlotGain(const ApuParameterRegisterWords& registerWords) {
	const ApuAudioSlot slot = registerWords[APU_PARAMETER_SLOT_INDEX] & APU_SLOT_INDEX_MASK;
	writeSlotRegisterWord(slot, APU_PARAMETER_GAIN_Q12_INDEX, registerWords[APU_PARAMETER_GAIN_Q12_INDEX]);
}

void ApuCommandExecutor::writeSlotRegisterWord(ApuAudioSlot slot, u32 parameterIndex, u32 word) {
	m_slots.writeRegisterWord(slot, parameterIndex, word);
	if ((m_slots.activeMask() & (1u << slot)) != 0u) {
		m_slots.loadRegisterWords(slot, m_slotRegisterDispatchWords);
		loadApuAudioSource(m_source, m_slotRegisterDispatchWords);
		if (apuParameterProgramsSourceBuffer(parameterIndex)) {
			if (!bindSource(m_source, m_memory.cartridgeController().selectedSlot())) {
				m_audioOutput.stopSlot(slot);
				m_activeSlots.deactivate(slot);
				m_fault.raise(APU_FAULT_SOURCE_RANGE, m_source.sourceAddr);
				return;
			}
			m_audioOutput.replaceVoiceSource(slot, m_source, m_sourceView, m_slotRegisterDispatchWords);
		} else {
			m_audioOutput.writeSlotRegisterWord(
				slot,
				m_source,
				m_slotRegisterDispatchWords,
				parameterIndex
			);
		}
	}
	m_selectedSlotLatch.refresh(m_commandRegisterWords[APU_PARAMETER_SLOT_INDEX] & APU_SLOT_INDEX_MASK);
}

bool ApuCommandExecutor::bindSource(const ApuAudioSource& source, u32 cartridgeSlot) {
	if (apuAudioSourceUsesGenerator(source)) {
		m_sourceView.bytes = {};
		m_sourceView.cartridgeSlot = cartridgeSlot;
		return true;
	}
	return m_sampleMemory.bindSource(source.sourceAddr, source.sourceBytes, cartridgeSlot, m_sourceView);
}

} // namespace bmsx

File: machine/ts/machine/devices/audio/command_executor.ts
import { IO_APU_SELECTED_SLOT_REG0, IO_APU_SLOT, IO_ARG_STRIDE } from '../../bus/io';
import type { Value } from '../../cpu/cpu';
import type { Memory } from '../../memory/memory';
import type { DeviceScheduler } from '../../scheduler/device';
import type { DeviceStatusLatch } from '../device_status';
import type { ApuActiveSlots } from './active_slots';
import type { ApuCommandFifo } from './command_fifo';
import type { ApuOutputMixer } from './output';
import { ApuSelectedSlotLatch } from './selected_slot_latch';
import type { ApuServiceClock } from './service_clock';
import {
	ApuSourceDma,
	apuParameterProgramsSourceBuffer,
	resolveApuAudioSource,
} from './source';
import type { ApuOutputVoiceState } from './save_state';
import type { ApuSlotBank } from './slot_bank';
import {
	APU_CMD_PLAY,
	APU_CMD_SET_SLOT_GAIN,
	APU_CMD_STOP_SLOT,
	APU_FAULT_BAD_CMD,
	APU_PARAMETER_FADE_SAMPLES_INDEX,
	APU_PARAMETER_GAIN_Q12_INDEX,
	APU_PARAMETER_REGISTER_COUNT,
	APU_PARAMETER_SLOT_INDEX,
	APU_SLOT_INDEX_MASK,
	APU_SLOT_PHASE_FADING,
	type ApuAudioSlot,
	type ApuAudioSource,
	type ApuParameterRegisterWords,
} from './contracts';

export class ApuCommandExecutor {
	private readonly commandDispatchRegisterWords = new Uint32Array(APU_PARAMETER_REGISTER_COUNT);
	private readonly slotRegisterDispatchWords = new Uint32Array(APU_PARAMETER_REGISTER_COUNT);

	public constructor(
		private readonly memory: Memory,
		private readonly audioOutput: ApuOutputMixer,
		private readonly scheduler: DeviceScheduler,
		private readonly commandFifo: ApuCommandFifo,
		private readonly sourceDma: ApuSourceDma,
		private readonly activeSlots: ApuActiveSlots,
		private readonly slots: ApuSlotBank,
		private readonly selectedSlotLatch: ApuSelectedSlotLatch,
		private readonly fault: DeviceStatusLatch,
		private readonly serviceClock: ApuServiceClock,
	) {}

	public drainCommandFifo(): void {
		while (!this.commandFifo.empty) {
			const command = this.commandFifo.popInto(this.commandDispatchRegisterWords);
			this.executeCommand(command, this.commandDispatchRegisterWords);
		}
	}

	public restoreOutputVoice(state: ApuOutputVoiceState): void {
		const slot = state.slot;
		const registerWords = this.slotRegisterDispatchWords;
		this.slots.loadRegisterWords(slot, registerWords);
		this.audioOutput.restoreVoice(
			slot,
			resolveApuAudioSource(registerWords),
			this.sourceDma.bytesForSlot(slot),
			registerWords,
			state,
		);
	}

	public static selectedSlotRegisterReadThunk(context: ApuCommandExecutor, addr: number): number {
		const nowCycles = context.scheduler.currentNowCycles();
		context.serviceClock.synchronize(nowCycles);
		const slot = context.memory.readIoU32(IO_APU_SLOT) & APU_SLOT_INDEX_MASK;
		const parameterIndex = (addr - IO_APU_SELECTED_SLOT_REG0) / IO_ARG_STRIDE;
		return context.slots.registerWord(slot, parameterIndex);
	}

	public static selectedSlotRegisterWriteThunk(context: ApuCommandExecutor, addr: number, value: Value): void {
		const nowCycles = context.scheduler.currentNowCycles();
		context.serviceClock.synchronize(nowCycles);
		const slot = context.memory.readIoU32(IO_APU_SLOT) & APU_SLOT_INDEX_MASK;
		context.writeSlotRegisterWord(slot, (addr - IO_APU_SELECTED_SLOT_REG0) / IO_ARG_STRIDE, (value as number) >>> 0);
		context.serviceClock.scheduleNext(nowCycles);
	}

	private executeCommand(command: number, registerWords: ApuParameterRegisterWords): void {
		switch (command) {
			case APU_CMD_PLAY:
				this.play(registerWords);
				return;
			case APU_CMD_STOP_SLOT:
				this.stopSlot(registerWords);
				return;
			case APU_CMD_SET_SLOT_GAIN:
				this.setSlotGain(registerWords);
				return;
			default:
				this.fault.raise(APU_FAULT_BAD_CMD, command);
				return;
		}
	}

	private play(registerWords: ApuParameterRegisterWords): void {
		const source = resolveApuAudioSource(registerWords);
		const slot = registerWords[APU_PARAMETER_SLOT_INDEX]! & APU_SLOT_INDEX_MASK;
		this.startPlay(source, slot, registerWords);
	}

	private startPlay(source: ApuAudioSource, slot: ApuAudioSlot, registerWords: ApuParameterRegisterWords): void {
		this.sourceDma.loadSlot(slot, source);
		this.activeSlots.setActive(slot, registerWords);
		this.audioOutput.playVoice(slot, source, this.sourceDma.bytesForSlot(slot), registerWords);
	}

	private stopSlot(registerWords: ApuParameterRegisterWords): void {
		const slot = registerWords[APU_PARAMETER_SLOT_INDEX]! & APU_SLOT_INDEX_MASK;
		const fadeSamples = registerWords[APU_PARAMETER_FADE_SAMPLES_INDEX]!;
		if ((this.slots.activeMask & (1 << slot)) === 0) {
			this.audioOutput.stopSlot(slot);
			this.activeSlots.stop(slot);
			return;
		}
		if (fadeSamples > 0) {
			this.activeSlots.setPhase(slot, APU_SLOT_PHASE_FADING);
			this.audioOutput.stopSlot(slot, fadeSamples);
			return;
		}
		this.audioOutput.stopSlot(slot);
		this.activeSlots.stop(slot);
	}

	private setSlotGain(registerWords: ApuParameterRegisterWords): void {
		const slot = registerWords[APU_PARAMETER_SLOT_INDEX]! & APU_SLOT_INDEX_MASK;
		this.writeSlotRegisterWord(slot, APU_PARAMETER_GAIN_Q12_INDEX, registerWords[APU_PARAMETER_GAIN_Q12_INDEX]!);
	}

	private writeSlotRegisterWord(slot: ApuAudioSlot, parameterIndex: number, word: number): void {
		this.slots.writeRegisterWord(slot, parameterIndex, word);
		if ((this.slots.activeMask & (1 << slot)) !== 0) {
			this.slots.loadRegisterWords(slot, this.slotRegisterDispatchWords);
			const source = resolveApuAudioSource(this.slotRegisterDispatchWords);
			if (apuParameterProgramsSourceBuffer(parameterIndex)) {
				this.sourceDma.loadSlot(slot, source);
				this.audioOutput.replaceVoiceSource(slot, source, this.sourceDma.bytesForSlot(slot), this.slotRegisterDispatchWords);
			} else {
				this.audioOutput.writeSlotRegisterWord(
					slot,
					source,
					this.slotRegisterDispatchWords,
					parameterIndex,
				);
			}
		}
		ApuSelectedSlotLatch.refreshThunk(this.selectedSlotLatch);
	}

}
